### PR TITLE
netlink communication improvments 

### PR DESCRIPTION
**Pull request by fousti**

### CODE REVIEW

File: lib/libnetlink/netlink.c [hunk 1]
```diff
@@ -390,53 +395,65 @@ int netl_listen(struct netl_handle *h, void *args)
...
-				hdr =
-					(struct nlmsghdr *) ((char *) hdr + NLMSG_ALIGN(len));
-			}
-
-			if (status) {
-				// content not read
-				return -1;
-			}
-		}
+    if (poll(fds, 1, NETL_POLL_TIMEOUT) > 0) {
```
Comment (baloo) [inline lib/libnetlink/netlink.c]: if poll returns -1 and errno == EINTR, you're expected to re-run poll here

Comment (baloo) [inline lib/libnetlink/netlink.c]: you're suppose to check errno, on return, everything else than errno EINTR is considered as fatal

File: lib/libnetlink/netlink.c [hunk 2]
```diff
@@ -390,53 +395,65 @@ int netl_listen(struct netl_handle *h, void *args)
...
+    
+      if (fds[0].revents & POLLIN) {
+        iov.iov_len = sizeof(buf);
+        status = recvmsg(h->fd, &msg, 0);
```
Comment (baloo) [inline lib/libnetlink/netlink.c]: I believe you're supposed to use `nl_recvmsgs_default(h)` here

File: lib/libnetlink/netlink.c [hunk 3]
```diff
@@ -4,6 +4,9 @@
 #include <linux/if_arp.h>
 #include <poll.h>
 
+#define ERR 4
+#define INFO 7
+#define DEBUG 8
```
Comment (baloo) [inline lib/libnetlink/netlink.c]: Do not redefine yours, use RTE_LOG_ERR, RTE_LOG_INFO, ... Or alias those near the `void (*log) (const char* msg, uint32_t lvl);` definition, maybe.

**Comment by baloo** [comment]
Actually libnl has a very nice notify framework, you should have a look.

``` C
static void
msg_parsed(struct nl_object *o, void *arg)
{
    struct rtnl_link *link = (struct rtnl_link *)o;
    state_t simple_state;

    switch (rtnl_link_get_operstate(link)) {
    case UP:
        simple_state = LINK_UP;
        break;
    case DOWN:
    case LOWERLAYERDOWN:
        simple_state = LINK_DOWN;
        break;
    default:
        simple_state = LINK_UNKNOWN;
        break;
    }

        // Do whatever stuff you want
}
static int
custom_handler(struct nl_msg *msg, void *arg)
{
    const struct nlmsghdr *msg_hdr;

    msg_hdr = nlmsg_hdr(msg);
    {
        switch (msg_hdr->nlmsg_type) {
        case RTM_NEWLINK:
            nl_msg_parse(msg, msg_parsed, arg);
            return 0;
        }
    }

    return 0;
}

struct nl_sock *nl_sk = nl_socket_alloc();
nl_connect(nl_sk, NETLINK_ROUTE);
nl_socket_disable_seq_check(nl_sk); // no sequence on notify events
nl_socket_modify_cb(nl_sk, NL_CB_VALID, NL_CB_CUSTOM,
                custom_handler, NULL);
nl_connect(nl_sk, NETLINK_ROUTE);
nl_socket_add_memberships(nl_sk, RTNLGRP_IPV4_IFADDR|RTNLGRP_IPV4_ROUTE, 0);
nl_socket_set_nonblocking(nl_sk);
fds.fd[0] = nl_socket_get_fd(nl_sk);
poll(fds, /*... */);
nl_recvmsgs_default(nl_sk);
```

If I were to rewrite the netlink part, I'd use this.
This sample has no error control, but it's just to give you an idea. Please have a look to https://github.com/baloo/leaf/blob/master/src/netlink.c for a more complete example, but you get the idea.

**Comment by fousti** [comment]
@baloo It feels like all this stuff could (should ?) be handle by an external library (parsing, validating netlink message and so on), this will allow us to focus only on processed the message and avoiding maintaining this logic.

Several options exists, libnl (https://www.infradead.org/~tgr/libnl/) seems complex and is split into several libraries and it's very "object-oriented", however it is worth to have a look at the implementation. 
libnetlink from iproute2 seems to be a good choice, but they recommend the use of libmnl (http://www.netfilter.org/projects/libmnl/) for new programs in their manual (man libnetlink).
@baloo @klnikita if i could have your feedback on those question it would be great.

File: lib/libnetlink/netlink.c [hunk 4]
```diff
@@ -392,49 +397,63 @@ int netl_listen(struct netl_handle *h, void *args)
 	while (h->closing != 1) {
 		int res = poll(fds, 1, NETL_POLL_TIMEOUT);
 		if (res < 0 && errno != EINTR) {
-			perror("error during cmdline_run poll");
-			return 0;
+			h->cb.log("Error while polling netlink socket", ERR);
```
Comment (baloo) [inline lib/libnetlink/netlink.c]: This does not exit right?

**Comment by baloo** [comment]
Well, libmnl is probably too minimalist for our use-case and having read the code, I'm pretty sure you would endup having to write a ton of code for parsing. And I'm not sure you want to go that way. But it's your call :)

**Comment by klnikita** [comment]
I have the same feeling as baloo for libmnl, switching to that will be counterproductive since anyway you will have to do the parsing yourself. The only thing it will bring is its one implementation for the polling/reading part without any benefit if its some more tuning to do to make it work at least the same way the current code do and an external dependency.

About libnl, it looks as a better choice than the previous one, at least it has a proper parser. If you want to try it, you can. Don't forget to check the default netlink socket settings that libnl will use and try to find if it has a message cache.

But, in any case, please first finish this pull request, it is almost finished, let's not drop it, it contains some improvement.

**Pull request closed without merging**